src/block/put.js

```diff
@@ -22,16 +23,36 @@ module.exports = (send) => {
       return callback(new Error('block.put accepts only one block'))
     }
 
+    let cid = opts.cid
+
     if (typeof block === 'object' && block.data) {
+      if (block.cid) cid = block.cid
       block = block.data
     }
 
-    sendOneFile(block, {}, (err, result) => {
+    let qs = Object.assign(opts, {
+      'input-enc': 'raw',
+      hashAlg: opts.mhtype || 'sha2-256'
+    })
+
+    if (cid && CID.isCID(cid)) {
+      qs.format = cid.codec
+      qs.hashAlg = multihashes.decode(cid.multihash).name
+    }
+    delete qs.cid
+
+    let _send = sendOneFile
+
+    if (qs.format && qs.format.startsWith('dag-')) {
+      _send = SendOneFile(send, 'dag/put')
+    }
```

Comment:
I'm mildly against adding this workaround here. I think it'll be unexpected for this method to call `dag/put` and without some comments or docs this behaviour will become confusing for contributors as well. It also sounds as though it's specific to go-ipfs and I'd rather not add implementation specific workarounds here in the API if we can avoid it. I'd like to know if js-ipfs has the same issue.

Is there an open issue in go-ipfs for this? It would be good to get this fixed there.